src/julia_to_gap.jl

```diff
@@ -93,148 +131,213 @@ function julia_to_gap(x::Rational{T}) where {T<:Integer}
-julia_to_gap(x::Char) = CharWithValue(Cuchar(x))
+GAP.@install GapObj(x::Char) = CharWithValue(Cuchar(x))
 
 ## Strings and symbols
-julia_to_gap(x::AbstractString) = MakeString(string(x))
-julia_to_gap(x::Symbol) = MakeString(string(x))
+GAP.@install GapObj(x::AbstractString) = MakeString(string(x))
+GAP.@install GapObj(x::Symbol) = MakeString(string(x))
+
+# helper functions for recursion
+function recursion_info(::Type{T}, obj, recursive::Bool, recursion_dict::GapCacheDict) where {T}
+    recursive && recursion_dict !== nothing && haskey(recursion_dict, obj) && return true, false, recursion_dict
+    rec = recursive && _needs_tracking_julia_to_gap(T)
+    if rec && recursion_dict === nothing
+        rec_dict = RecDict()
+    else
+        rec_dict = recursion_dict
+    end
+
+    return false, rec, rec_dict
+end
 
-## Generic caller for optional arguments
-julia_to_gap(obj::Any; recursive::Bool) = julia_to_gap(obj)
-julia_to_gap(obj::Any, recursion_dict::IdDict{Any,Any}; recursive::Bool = true) = julia_to_gap(obj)
+function handle_recursion(obj, ret_val, rec::Bool, rec_dict::GapCacheDict)
+    if rec_dict !== nothing
+        rec_dict[obj] = ret_val
+    end
+    return rec ? rec_dict : nothing
+end
 
 ## Arrays (including BitVector)
-function julia_to_gap(
+function GapObj_internal(
     obj::AbstractVector{T},
-    recursion_dict::IdDict{Any,Any} = IdDict();
-    recursive::Bool = false,
-) where {T}
+    recursion_dict::GapCacheDict,
+    ::Val{recursive},
+) where {T, recursive}
+
+    ret, rec, rec_dict = recursion_info(T, obj, recursive, recursion_dict)
+    ret && return rec_dict[obj]
 
     len = length(obj)
     ret_val = NewPlist(len)
-    if recursive
-        recursion_dict[obj] = ret_val
-    end
+
+    recursion_dict = handle_recursion(obj, ret_val, rec, rec_dict)
+
+    # Set the subobjects.
     for i = 1:len
         x = obj[i]
         if x === nothing
             continue
         end
         if recursive
-            x = get!(recursion_dict, x) do
-                julia_to_gap(x, recursion_dict; recursive)
-            end
+            # Convert the subobjects.
+            # Do not care about findíng them in the dictionary
+            # or adding them to the dictionary,
+            # since their conversion method decides about that.
+            res = GAP.GapObj_internal(x, recursion_dict, Val(true))
+        else
+            res = x
+        end
+        ret_val[i] = res
+    end
+
+    return ret_val
+end
+
+## Sets
+function GapObj_internal(
+    obj::Set{T},
+    recursion_dict::GapCacheDict,
+    ::Val{recursive},
+) where {T, recursive}
+
+    ret, rec, rec_dict = recursion_info(T, obj, recursive, recursion_dict)
+    ret && return rec_dict[obj]
+
+    ret_val = NewPlist(length(obj))
+
+    recursion_dict = handle_recursion(obj, ret_val, rec, rec_dict)
+
+    for x in obj
+        if recursive
+            res = GapObj_internal(x, recursion_dict, Val(true))
+        else
+            res = x
         end
```

Comment:
Here and elsewhere, might also write
```suggestion
        res = recursive ? GapObj_internal(x, recursion_dict, Val(true)) : x
```
but this, too, can wait for a future PR (or we just don't do it)